Modules/_datetimemodule.c

```diff
@@ -47,39 +48,150 @@ typedef struct {
+#define INTERP_KEY "cached-datetime-module"
+
+static PyObject *
+get_current_module(PyInterpreterState *interp)
+{
+    PyObject *dict = PyInterpreterState_GetDict(interp);
+    if (dict == NULL) {
+        return NULL;
+    }
+    PyObject *ref = NULL;
+    if (PyDict_GetItemStringRef(dict, INTERP_KEY, &ref) < 0) {
+        return NULL;
+    }
+    if (ref == NULL) {
+        return NULL;
+    }
+    PyObject *mod = NULL;
+    (void)PyWeakref_GetRef(ref, &mod);
+    Py_DECREF(ref);
+    return mod;
+}
+
+#define GET_CURRENT_STATE(ST_VAR)                               \
+    NULL;                                                       \
+    PyObject *current_mod = NULL;                               \
+    do {                                                        \
+        PyInterpreterState *interp = PyInterpreterState_Get();  \
+        current_mod = get_current_module(interp);               \
+        assert(current_mod != NULL);                            \
+        ST_VAR = get_module_state(current_mod);                 \
+    } while (0)
+#define RELEASE_CURRENT_STATE(ST_VAR)   \
+    Py_DECREF(current_mod)
 
-#define PyTime_Check(op) PyObject_TypeCheck(op, get_datetime_state()->time_type)
-#define PyTime_CheckExact(op) Py_IS_TYPE(op, get_datetime_state()->time_type)
+static int
+set_current_module(PyInterpreterState *interp, PyObject *mod)
+{
+    assert(mod != NULL);
+    PyObject *dict = PyInterpreterState_GetDict(interp);
+    if (dict == NULL) {
+        return -1;
+    }
+    PyObject *ref = PyWeakref_NewRef(mod, NULL);
+    if (ref == NULL) {
+        return -1;
+    }
+    int rc = PyDict_SetItemString(dict, INTERP_KEY, ref);
+    Py_DECREF(ref);
+    return rc;
+}
 
-#define PyDelta_Check(op) PyObject_TypeCheck(op, get_datetime_state()->delta_type)
-#define PyDelta_CheckExact(op) Py_IS_TYPE(op, get_datetime_state()->delta_type)
+static void
+clear_current_module(PyInterpreterState *interp, PyObject *expected)
+{
+    PyObject *exc = PyErr_GetRaisedException();
 
-#define PyTZInfo_Check(op) PyObject_TypeCheck(op, get_datetime_state()->tzinfo_type)
-#define PyTZInfo_CheckExact(op) Py_IS_TYPE(op, get_datetime_state()->tzinfo_type)
+    PyObject *current = NULL;
+
+    PyObject *dict = PyInterpreterState_GetDict(interp);
+    if (dict == NULL) {
+        goto error;
+    }
+
+    if (expected != NULL) {
+        PyObject *ref = NULL;
+        if (PyDict_GetItemStringRef(dict, INTERP_KEY, &ref) < 0) {
+            goto error;
+        }
+        int rc = PyWeakref_GetRef(ref, &current);
+        Py_DECREF(ref);
+        if (rc < 0) {
+            goto error;
+        }
+        if (current != expected) {
+            goto finally;
+        }
+    }
+
+    if (PyDict_DelItemString(dict, INTERP_KEY) < 0) {
+        if (!PyErr_ExceptionMatches(PyExc_KeyError)) {
+            goto error;
+        }
+    }
+
+    goto finally;
+
+error:
+    PyErr_Print();
+
+finally:
+    Py_XDECREF(current);
+    PyErr_SetRaisedException(exc);
+}
```

Comment:
This is the implementation for tracking the "current" module, to facilitate looking up module state from type methods.

The `GET_CURRENT_STATE()` and `RELEASE_CURRENT_STATE()` macros are used in the method implementations.  `get_current_module()` and `set_current_module()` are used in the module exec function.  `clear_current_module()` is used in the module clear/free functions.